src/libraries/Common/src/System/Net/Http/Http3/QPack/QPackEncoder.cs

```diff
@@ -0,0 +1,424 @@
+        {
+            if (EncodeNameString(name, destination, out int nameLength) && EncodeValueString(values, valueSeparator, destination.Slice(nameLength), out int valueLength))
+            {
+                bytesWritten = nameLength + valueLength;
+                return true;
+            }
+
+            bytesWritten = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Encodes just the value part of a Literawl Header Field Without Static Name Reference. Must call <see cref="EncodeValueString(string, Span{byte}, out int)"/> after to encode the header's value.
+        /// </summary>
+        public static byte[] EncodeLiteralHeaderFieldWithoutNameReferenceToArray(string name)
+        {
+            Span<byte> temp = name.Length < 256 ? stackalloc byte[256 + IntegerEncoder.MaxInt32EncodedLength] : new byte[name.Length + IntegerEncoder.MaxInt32EncodedLength];
+
+            bool res = EncodeNameString(name, temp, out int nameLength);
+            Debug.Assert(res == true);
+
+            return temp.Slice(0, nameLength).ToArray();
+        }
+
+        public static byte[] EncodeLiteralHeaderFieldWithoutNameReferenceToArray(string name, string value)
+        {
+            Span<byte> temp = (name.Length + value.Length) < 256 ? stackalloc byte[256 + IntegerEncoder.MaxInt32EncodedLength * 2] : new byte[name.Length + value.Length + IntegerEncoder.MaxInt32EncodedLength * 2];
+
+            bool res = EncodeLiteralHeaderFieldWithoutNameReference(name, value, temp, out int bytesWritten);
+            Debug.Assert(res == true);
+
+            return temp.Slice(0, bytesWritten).ToArray();
+        }
+
+        private static bool EncodeValueString(string s, Span<byte> buffer, out int length)
+        {
+            if (buffer.Length != 0)
+            {
+                buffer[0] = 0;
+                if (IntegerEncoder.Encode(s.Length, 7, buffer, out int nameLength))
+                {
+                    buffer = buffer.Slice(nameLength);
+                    if (buffer.Length >= s.Length)
+                    {
+                        EncodeValueStringPart(s, buffer);
+
+                        length = nameLength + s.Length;
+                        return true;
+                    }
+                }
+            }
+
+            length = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Encodes a value by concatenating a collection of strings, separated by a separator string.
+        /// </summary>
+        public static bool EncodeValueString(ReadOnlySpan<string> values, string separator, Span<byte> buffer, out int length)
+        {
+            if (values.Length == 1)
+            {
+                return EncodeValueString(values[0], buffer, out length);
+            }
+
+            if (values.Length == 0)
+            {
+                return EncodeValueString(string.Empty, buffer, out length);
```

Comment:
Can/should we have a precomputed value for this that we can just copy into the buffer?

Reply:
> Or is it cheap enough / this rare enough that it's not worth it?

This branch will only be hit if HttpHeaderCollection's array of values can be zero-length. Which may not be possible -- I need to check. Might be able to just assert this.